app/controllers/leader_sessions_controller.rb

```diff
@@ -0,0 +1,30 @@
+class LeaderSessionsController < ApplicationController
+  skip_before_action(:force_leader_sign_in, { :only => [:registration_form, :session_form, :add_cookie] })
+
+  def session_form
+    render({ :template => "leader_sessions/sign_in.html.erb" })
+  end
+
+  def registration_form
+    render({ :template => "leader_sessions/sign_up.html.erb" })
+  end
+
+  def add_cookie
+    leader = Leader.where({ :email => params.fetch(:email) }).at(0)
+
+    if leader != nil && leader.authenticate(params.fetch(:password))
```

Comment:
Despite the best practice of not revealing whether it was email or password that was incorrect, I think it would be clearer for students to handle the two cases separately with nested `if` statements and different `notice`s. See the Codio workspace for an example:

https://codio.com/dstudent43/photogram-signin